client/alloc_runner.go

```diff
@@ -856,26 +849,37 @@ func (r *AllocRunner) SetPreviousAllocDir(allocDir *allocdir.AllocDir) {
 // destroyTaskRunners destroys the task runners, waits for them to terminate and
 // then saves state.
 func (r *AllocRunner) destroyTaskRunners(destroyEvent *structs.TaskEvent) {
-	runners := r.getTaskRunners()
-
-	// First destroy the leader
-	for _, tr := range runners {
-		if tr.task.Leader {
-			r.logger.Printf("[DEBUG] client: destroying leader task %q of task group %q first", tr.task.Name, tr.alloc.TaskGroup)
-			tr.Destroy(destroyEvent)
-			<-tr.WaitCh()
+	// First destroy the leader if one exists
```

Comment:
https://github.com/hashicorp/nomad/pull/2888/commits/2569c58cb7b6c0bf366d777358a9ce87966583f6

AllocRunner should not directly access `TaskRunner.task` fields because TaskRunner may be mutating task concurrently at any time. (It's fine for AllocRunner to call _methods_ on TaskRunner because those can do locking to be threadsafe).

So this commit changed leader detection to look at `AllocRunner.alloc` instead of peeking into `TaskRunner.task`.